src/note.rs

```diff
@@ -108,18 +112,33 @@ impl PartialEq for Note {
 impl Eq for Note {}
 
 impl Note {
-    pub(crate) fn from_parts(
+    /// Creates a `Note` from its component parts.
+    ///
+    /// Returns `None` if a valid [`NoteCommitment`] cannot be derived from the note.
+    ///
+    /// # Caveats
+    ///
+    /// This low-level constructor enforces that the provided arguments produce an
+    /// internally valid `Note`. However, it allows notes to be constructed in a way that
+    /// violates required security checks for note decryption, as specified in
+    /// [Section 4.19] of the Zcash Protocol Specification. Users of this constructor
+    /// should only call it with note components that have been fully validated by
+    /// decrypting a received note according to [Section 4.19].
+    ///
+    /// [Section 4.19]: https://zips.z.cash/protocol/protocol.pdf#saplingandorchardinband
+    pub fn from_parts(
         recipient: Address,
         value: NoteValue,
         rho: Nullifier,
         rseed: RandomSeed,
```

Comment:
It occurs to me now that a potentially simpler API for downstream users would be for the public `from_parts` to take `rseed` as a `[u8; 32]` and not expose the `RandomSeed` type externally at all (currently the only thing you can do with it publicly is wrap it around a byte array, and I don't see this changing in future). The `RandomSeed::from_bytes` check would then be done internally to `Note::from_parts` as:
```rust
impl Note {
    pub fn from_parts(
        recipient: Address,
        value: NoteValue,
        rho: Nullifier,
        rseed: [u8; 32],
    ) -> CtOption<Self> {
        RandomSeed::from_bytes(rseed, &rho)
            .and_then(|rseed| Self::from_parts_internal(recipient, value, rho, rseed))
    }

    fn from_parts_internal(
        recipient: Address,
        value: NoteValue,
        rho: Nullifier,
        rseed: RandomSeed,
    ) -> CtOption<Self> {
        ...
    }

    pub fn rseed(&self) -> &[u8; 32] {
        self.rseed.as_bytes()
    }
}
```

On the other hand, having the `RandomSeed` type exist may make it easier for crate users to keep it typed in the space between where they interact with their database and where they interact with `Note`. Then again, there is nothing stopping them only using `RandomSeed` immediately before `Note::from_parts`, which makes the type safety completely pointless in their usage:
```rust
let note = RandomSeed::from_bytes(rseed, &rho)
    .and_then(|rseed| Note::from_parts(recipient, value, rho, rseed))
````

Reply:
Can we call the local variable name on the line that starts with `.and_then` something other than `rseed`?

Reply:
I think that this is okay as it is currently; it makes the role of `rho` clear in constructing the `RandomSeed` value.

Reply:
The `RandomSeed` being constructed beforehand makes it more obvious to the consumer whether the note construction went wrong due to a mismatched rho/rseed, or something else. I don't know how to weigh that against the ergonomics of it.